### PR TITLE
Initial commit of Statestore APIs (GetState, SaveState)

### PR DESCRIPTION
**Pull request by chinzhiweiblank**
# Description

_Please explain the changes you've made_
Added get and save state. Implementation for etag, metadata and state options are not ready yet.
## Issue reference

We strive to have all PR being opened based on an issue, where the problem or feature have been discussed prior to implementation.

Please reference the issue this PR will close: #74 

## Checklist

Please make sure you've completed the relevant tasks for this PR, out of the following list:

* [ ] Code compiles correctly
* [ ] Created/updated tests
* [ ] Extended the documentation

### CODE REVIEW

File: dapr/clients/grpc/_state.py [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+class State:
+    def __init__(self, key, value, etag, options: StateOptions = None):
+        self.key = key
+        self.value = value
+        self.etag = etag
+        self.options = options
+
+
+class Consistency(Enum):
+    EVENTUAL = common_v1.StateOptions.StateConsistency.CONSISTENCY_EVENTUAL
```
Comment (youngbupark) [inline dapr/clients/grpc/_state.py:14]: please use snake_case for enum items.

File: dapr/clients/grpc/client.py [hunk 2]
```diff
@@ -284,6 +284,56 @@ def publish_event(
...
+            resp = self._stub.GetState(req)
+            return resp
+        except Exception as e:
+            return e
+
+    def save_state(
+        self,
+        store_name: str,
+        key: str,
+        value: object,
```
Comment (youngbupark) [inline dapr/clients/grpc/client.py:305]: value: Union[str, bytes]

File: dapr/clients/grpc/client.py [hunk 3]
```diff
@@ -284,6 +284,56 @@ def publish_event(
...
+        etag: str = "",
+        state_options: common_v1.StateOptions = None
+    ):
```
Comment (youngbupark) [inline dapr/clients/grpc/client.py:308]: add metadata arg as optional. please see the other implementation.

File: dapr/clients/grpc/client.py [hunk 4]
```diff
@@ -284,6 +284,56 @@ def publish_event(
...
+            state = [common_v1.StateItem(
+                key=state['key'], value=value.encode('utf-8'), etag=etag.encode('utf-8'),
+                options=state_options)]
+            if state_options is not None:
+                state.options = state_options
+            req = api_v1.SaveStateRequest(store_name=store_name, states=state)
+            response = self._stub.SaveState(req)
+            return response
+        except Exception as e:
+            return e
```
Comment (youngbupark) [inline dapr/clients/grpc/client.py:335]: why it returns exception?

File: dapr/clients/grpc/client.py [hunk 5]
```diff
@@ -284,6 +284,56 @@ def publish_event(
...
+            if len(store_name) == 0 or len(store_name.strip()) == 0:
+                raise ValueError("State store name cannot be empty")
+            state = [common_v1.StateItem(
+                key=state['key'], value=value.encode('utf-8'), etag=etag.encode('utf-8'),
+                options=state_options)]
+            if state_options is not None:
+                state.options = state_options
+            req = api_v1.SaveStateRequest(store_name=store_name, states=state)
+            response = self._stub.SaveState(req)
+            return response
```
Comment (youngbupark) [inline dapr/clients/grpc/client.py:333]: all api needs to return DaprResponse object which includes response headers from dapr runtime.

Comment (chinzhiweiblank) [reply dapr/clients/grpc/client.py:333]: I noticed for `publish_event`, it's
```
        response, call = self._stub.PublishEvent.with_call(req, metadata=metadata)

        return DaprResponse(
            headers=call.initial_metadata(),
            trailers=call.trailing_metadata())
```
Do I implement
```
            req = api_v1.SaveStateRequest(store_name=store_name, states=state)
            response, call = self._stub.SaveState.with_call(req, metadata=metadata)
        return DaprResponse(
            headers=call.initial_metadata(),
            trailers=call.trailing_metadata())

```
?

Comment (youngbupark) [reply dapr/clients/grpc/client.py]: yes you need to do. I merged my pr. please check the latest code. I removed trailer part from response.

File: dapr/clients/grpc/_state.py [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+    def __init__(self, threshold, interval, pattern: RetryPattern):
+        self.threshold = threshold
+        self.interval = interval
+        if pattern is None:
+            self.pattern = common_v1.StateOptions.RetryPattern.RETRY_UNSPECIFIED
+
+
+class RetryPattern(Enum):
+    linear = common_v1.StateOptions.StateRetryPolicy.RetryPattern.RETRY_LINEAR
+    exponential = common_v1.StateOptions.StateRetryPolicy.RetryPattern.RETRY_EXPONENTIAL
```
Comment (youngbupark) [inline dapr/clients/grpc/_state.py:33]: both retry related protobuf msgs will be deprecated in this milestone. please move RetryPolicy and retryPattern.

File: dapr/clients/grpc/client.py [hunk 7]
```diff
@@ -284,6 +284,61 @@ def publish_event(
...
+    def get_state(
+            self,
+            store_name: str,
+            key: str,
+            metadata: Optional[MetadataTuple] = ()) -> DaprResponse:
+        if len(store_name) == 0 or len(store_name.strip()) == 0:
+            raise ValueError("State store name cannot be empty")
+        req = api_v1.GetStateRequest(store_name=store_name, key=key, metadata=metadata)
+        response, call = self._stub.GetState.with_call(req)
+        return DaprResponse(
```
Comment (youngbupark) [inline dapr/clients/grpc/client.py:296]: get_state response must have data. please define `StateResponse` by referring to https://github.com/dapr/python-sdk/blob/c592371ba38f056487be519ccf8d6c1a42a75737/dapr/clients/grpc/_response.py#L196

File: dapr/clients/grpc/client.py [hunk 8]
```diff
@@ -284,6 +284,61 @@ def publish_event(
...
+            headers=call.initial_metadata(),
+            trailers=call.trailing_metadata())
+
+    def save_state(
+            self,
+            store_name: str,
+            key: str,
+            value: Union[bytes, str],
+            etag: str = "",
```
Comment (youngbupark) [inline dapr/clients/grpc/client.py:305]: Please Use `Optional`
etag: Optional[str] = None,

File: dapr/clients/grpc/client.py [hunk 9]
```diff
@@ -284,6 +284,61 @@ def publish_event(
...
+            state_options: StateOptions = None,
```
Comment (youngbupark) [inline dapr/clients/grpc/client.py:306]: Please use `Optional`

state_options: Optional[StateOptionas] = None

File: tests/clients/fake_dapr_server.py [hunk 10]
```diff
@@ -74,6 +76,21 @@ def PublishEvent(self, request, context):
         context.set_trailing_metadata(trailers)
         return empty_pb2.Empty()
 
+    def SaveState(self, request, context):
+        headers = ()
+        trailers = ()
+        for state in request.states:
+            state = json.loads(MessageToJson(state))
+            self.store[state['key']] = state['value'].decode('utf-8')
```
Comment (youngbupark) [inline tests/clients/fake_dapr_server.py:84]: we have helpers to convert str to bytes or vice versa. - https://github.com/dapr/python-sdk/blob/c592371ba38f056487be519ccf8d6c1a42a75737/dapr/clients/grpc/_helpers.py#L55-L72

File: dapr/clients/grpc/_state.py (hunk 6 above)
Comment (chinzhiweiblank) [reply dapr/clients/grpc/_state.py]: `dapr/clients/grpc/_state.py:14: error: "GeneratedProtocolMessageType" has no attribute "StateConsistency"  [attr-defined]` 
I keep getting type errors when I try to import them. Is there a way to resolve it?
```
class Consistency(Enum):
    eventual = common_v1.StateOptions.StateConsistency.CONSISTENCY_EVENTUAL
    strong = common_v1.StateOptions.StateConsistency.CONSISTENCY_STRONG
```

Comment (youngbupark) [reply dapr/clients/grpc/_state.py]: please ignore typing error by adding `# type: ignore` at the end of line.

**Comment by youngbupark** [comment]
@mukundansundar @pruthvidhodda can you please review this pr ?

File: dapr/clients/grpc/client.py [hunk 11]
```diff
@@ -297,6 +299,61 @@ def publish_event(
...
+            self,
+            store_name: str,
+            key: str,
+            metadata: Optional[MetadataTuple] = ()) -> DaprResponse:
+        if len(store_name) == 0 or len(store_name.strip()) == 0:
+            raise ValueError("State store name cannot be empty")
+        req = api_v1.GetStateRequest(store_name=store_name, key=key)
+        response, call = self._stub.GetState.with_call(req, metadata=metadata)
+        return StateResponse(
+            data=to_bytes(response.data),
```
Comment (youngbupark) [inline dapr/clients/grpc/client.py:312]: response.data must be bytes. we do not need to convert it.

**Comment by youngbupark** [comment]
@chinzhiweiblank overall looks good to me. @pruthvidhodda please review this pr.

**Comment by chinzhiweiblank** [comment]
@youngbupark  I’m not done testing the state options yet. It will take a day or two.

**Comment by youngbupark** [comment]
> @youngbupark I’m not done testing the state options yet. It will take a day or two.

Got it. Thanks for all your contribution! let's aim at deliver your feature in this milestone. we will help you to complete this PR.

File: dapr/clients/grpc/client.py [hunk 12]
```diff
@@ -297,6 +299,61 @@ def publish_event(
...
+                    store_name='state_store'
+                    states=[{'key': 'key1', 'value': 'value1'}],
+                )
+        Args:
+            store_name (str): the state store name to save to
+            states (List[dict]): the key-value pairs to be saved
+        Returns:
+            None
+        """
+        if not isinstance(value, bytes) and not isinstance(value, str):
```
Comment (youngbupark) [inline dapr/clients/grpc/client.py:337]: you can make it shorter:
```
if not isinstance(value, (str, bytes)): 
```

**Comment by youngbupark** [comment]
> > @youngbupark I’m not done testing the state options yet. It will take a day or two.
> 
> Got it. Thanks for all your contribution! let's aim at deliver your feature in this milestone. we will help you to complete this PR.

@chinzhiweiblank can we merge this one now? I need to use your change today :) you can validate it later and fix the bug as a separate pr.

File: tests/clients/test_dapr_grpc_client.py [hunk 13]
```diff
@@ -165,6 +166,19 @@ def test_dapr_api_token_insertion(self):
...
+        value = "value_1"
+        resp = dapr.save_state(
+            store_name="statestore",
+            key=key,
+            value=value
+        )
+        resp = dapr.get_state(store_name="statestore", key=key)
+
+        self.assertEqual(to_bytes(value), resp.data)
+
```
Comment (youngbupark) [inline tests/clients/test_dapr_grpc_client.py:181]: Please add negative test for the arguments if possible

File: dapr/clients/grpc/_state.py [hunk 14]
```diff
@@ -0,0 +1,35 @@
...
+
+
+class Concurrency(Enum):
+    first_write = common_v1.StateOptions.StateConcurrency.CONCURRENCY_FIRST_WRITE  # type: ignore
+    last_write = common_v1.StateOptions.StateConcurrency.CONCURRENCY_LAST_WRITE  # type: ignore
+
+
+class StateOptions:
+    def __init__(self, consistency=None,
+                 concurrency=None, retry_policy=None):
```
Comment (youngbupark) [inline dapr/clients/grpc/_state.py:25]: def __init__(self, consistency=None,
                 concurrency=None, retry_policy=None):

Use Optional

def __init__(self, Optional[consistency]=None,
                 Optional[concurrency]=None, Optional[retry_policy]=None):

File: dapr/clients/grpc/_state.py [hunk 15]
```diff
@@ -0,0 +1,35 @@
+from enum import Enum
+from dapr.proto import common_v1
+
+
+class State:
+    def __init__(self, key, value, etag, options):
```
Comment (youngbupark) [inline dapr/clients/grpc/_state.py:6]: please add comments for each class and enum

**Pull request merged**